data-picker

IlyaKirdun opened the pull request:
Переделал способ обновления страницы в button.spec,dynamicProperties.spec,links.spec.ts;
Добавил тип MonthList в types.ts;
Мелкие исправления;
Добавил dataPicker.page;dataPicker.spec(12 кейсов)

12 тестов для страницы "Date Picker":
Проверка даты из поля виджета Select date;
Проверяем функционал ручного ввода даты в "Select Date";
Проверка функционала выбора месяца и года через селект меню в "Select Date";
Проверяем пагинацию дней месяца в "Select Date";
Проверяем выбор полной даты в "Select Date";
Проверка даты из поля виджета "Data And Time";
Проверяем функционал выбора месяца и года через выпадающие меню виджета "Date And Time";
Проверяем пагинацию дней месяца виджета "Data And Time";
Проверка выбора дня месяца виджета "Date And Time";
Проверка функционала выбора даты виджета "Data And Time";
Проверка ручного ввода даты виджета "Date And Time";
Проверка функционала выпадающего меню года в "Date And Time";

IlyaKirdun committed:
Добавил файл dataPicker.page
Добавил тип MonthList в types

IlyaKirdun committed:
Добавил метод pickSelectMenuByName();
Переделал метод pickDropdownMenuByName();
Удалил неиспользуемые локаторы и методы: selectYearInDropdownMenu(), selectMonthInDropdownMenu().

IlyaKirdun committed:
Переработал обновление страницы в: button.spec;links.spec;dynamicProperties.spec .
Переработал методы и интегрировал кастомный тип MonthList в: selectMonthOrYearInSelectMenu();selectMonthOrYearInDropdownMenu() в dataPicker.page .
Переработал метод navigateToMainPage() .
Мини-фикс в методе removeAds() в functions .

IlyaKirdun committed:
Добавил dataPicker.spec;
Добавил метод verifyMonthOrYearInSelectMenu(),verifyMonthOrYearInDropdownMenu(),clickEnterOnKeyboard();
Переделал метод verifyDayColor(),selectMonthOrYearInDropdownMenu().

IlyaKirdun committed:
Добавил dataPicker.spec;
Добавил методы getCurrentMonthOrYearInDropdownMenu(),getCurrentMonthOrYearInSelectMenu(),verifyTimeColor(),clickMonthOrYearDropdownMenu(),selectTimeByNumber();
Переработал методы selectDayByNumber(),clickEnterOnKeyboard();
Незначительные исправления.

IlyaKirdun committed:
Добавил dataPicker.spec;
Добавил методы getCurrentMonthOrYearInDropdownMenu(),getCurrentMonthOrYearInSelectMenu(),verifyTimeColor(),clickMonthOrYearDropdownMenu(),selectTimeByNumber();
Переработал методы selectDayByNumber(),clickEnterOnKeyboard();
Незначительные исправления.

IlyaKirdun committed:
Merge remote-tracking branch 'origin/data-picker' into data-picker

IlyaKirdun committed:
Вернул хэдлес

rybignat commented on .idea/dictionaries/project.xml:
Это еще что?)

rybignat commented on .idea/inspectionProfiles/Project_Default.xml:
Что это такое?)

rybignat commented on page/main.page.ts:
```diff
@@ -9,28 +9,41 @@ export default class MainPage {
-        return this.page
+    const maxAttempts = 5;
+
+    for (let attempt = 0; attempt < maxAttempts; attempt++) {
```
отвыкай от конструкции "н++", некоторыми линтерами она запрещается, лучше "н += 1"

rybignat replied:
Не исправил и зарезолвил, не хорошо.
Резолвить должен я, после того как увижу исправления

rybignat commented on page/main.page.ts:
```diff
@@ -9,28 +9,41 @@ export default class MainPage {
+  private async attemptNavigate(page: Page, url: string): Promise<boolean> {
+    const response = await page.goto(url, { waitUntil: 'domcontentloaded' });
+    return response !== null && !this.isErrorCode(response.status());
+  }
```
приватные функции располагаются сверху класса, над публичными методами

rybignat commented on page/main.page.ts:
```diff
@@ -9,28 +9,41 @@ export default class MainPage {
+            return this.page;
+          }
+        } finally {
+          // await newPage.close();
```
Это что?)

rybignat commented on page/main.page.ts:
```diff
@@ -9,28 +9,41 @@ export default class MainPage {
+        }
+
+        // Вторая попытка на новой странице
+        const newPage = await this.page.context().newPage();
```
По идее вот эта часть должна быть в catch блоке

IlyaKirdun replied:
catch внутри try??? Звучит странно.

rybignat commented on page/main.page.ts:
Пересмотреть архитектуру

IlyaKirdun replied:
Хотелось бы хоть какой то конкретики.

rybignat commented on utils/components/dataPicker.page.ts:
```diff
@@ -0,0 +1,110 @@
+
+  getCurrentDateForSelectDate() {
+    const date = new Date()
+
+    return date.toLocaleDateString("en-US", { year: 'numeric', month: '2-digit', day: '2-digit' })
+  }
```
return new Date().toLocaleDateString("en-US", { year: 'numeric', month: '2-digit', day: '2-digit' })

rybignat commented on utils/components/dataPicker.page.ts:
```diff
@@ -0,0 +1,110 @@
+    await expect(this.page.locator(`//select[contains(@class,"${menuName}")]`)).toHaveValue(valueMonthOrYear)
+  }
+
+  async verifyMonthOrYearInDropdownMenu(menuName: 'month' | 'year', value: MonthList | number | string | null){
```
У тебя и 4 типа данных на один аргумент. Как то многовато, не находишь? И зачем там вообще налл?

rybignat commented on utils/components/dataPicker.page.ts:
Проставить типы переменных, там где они отсутствуют

rybignat replied:
Раз уж начал писать приватные методы, пиши вспомогательные приваты типа click`a

IlyaKirdun replied:
По вспомогательным приватам не совсем понимаю что и как сделать

rybignat replied:
вместо того что бы каждый раз писать this.page.locator....
Можно написать приватный метод click, который будет в себя принимать селектор и опции для клика.

IlyaKirdun replied:
Но ведь в спеке у меня клики применяются только на динамических Xpath, каким образом я напишу метод под разные пути с разными ключевыми словами???

rybignat replied:
Ты не понял. Это метод вместо this.page.locator(...).click()
Он берет в себя селектор, динимачейский, хуипампический, главное что бы рабочий, и кликает на него.

rybignat commented on utils/functions.ts:
```diff
@@ -23,8 +23,8 @@ export async function removeAds(page: Page): Promise<void> {
-    actual: string | number,
-    expected: string | number,
+    actual: string | number | null,
+    expected: string | number | null,
```
Ты собираешься куда то передавать налл?

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+import DatePicker from "../../utils/components/dataPicker.page";
+import {assertByState, removeAds} from "../../utils/functions";
+import NavigationBar from "../../utils/components/navigationBar";
+import {MonthList} from "../../utils/types";
```
переделать под относительные пути

IlyaKirdun replied:
Ты может хотел сказать переделать под абсолютные, ведь тут и так относительные?)

rybignat replied:
Немного спутался, не относительные а ассоциативные, они же allias

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    navigationBar = new NavigationBar(page)
+    datePicker = new DatePicker(page)
+
+    await removeAds(page)
+    await mainPage.clickOnElement('Widgets')
+    await navigationBar.clickElementInNavigationBar('Date Picker')
+    await removeAds(page)
```
добавить декоратор для описания

IlyaKirdun replied:
Почитал про декораторы и от слова совсем не понял как их тут применить.

rybignat replied:
бл, в степы оберни

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    const day: number = 1
+    const stringMonth: MonthList = 'January'
+    const numberMonth: number = 1
+    const year: string = '2011'
```
обернуть в объект, вынести в дискрайб. Шесть констант это как то слишком, не находишь?

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    })
+  })
+
+  test('CASE_2: Проверяем функционал ручного ввода даты в "Select Date".', async () => {
```
Привыкай разбирать тесты на относительно глобальные декораторы типа Pre-conditions, Testing, Prost-conditions. При больших объемах теста, упростит навигацию

IlyaKirdun replied:
Так ведь этот тест считается полным, типо от начала до конца.

rybignat replied:
Все что до очистви поля ввода включительно можно считать прекондами это раз
Это распространяется не только на этот конкретный тест, а на остальные в том числе, это два

rybignat commented on utils/components/dataPicker.page.ts:
```diff
@@ -0,0 +1,110 @@
+    const currentColor: string = await element.evaluate(el => getComputedStyle(el).backgroundColor)
+
+    expect(currentColor).toBe(color)
+  }
```
Переименовать, ты проверяешь в первую очередь состояние, а цвет как индикатор

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+
+  test('CASE_4: Проверяем пагинацию дней месяца в "Select Date".', async () => {
+    let initialMonth: string = ''
+    let currentMonth: string = ''
```
если у тебя указан тип данных, зачем там пустая срока?

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+
+    await test.step(`Сравниваем текущий месяц с ${initialMonth}.`, async () => {
+      currentMonth = await datePicker.getCurrentMonthOrYearInSelectMenu('month')
+      await assertByState(initialMonth, currentMonth, 'notMatch')
```
зачем тебе лишний раз сохранять, если больше нигде не используешь?
await assertByState(initialMonth, await datePicker.getCurrentMonthOrYearInSelectMenu('month'), 'notMatch')

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    })
+
+    await test.step('Проверяем корректное отображение года в селект меню "Year".', async () => {
+      await datePicker.verifyMonthOrYearInSelectMenu("year", `${year}`)
```
двойные кавычки.
Напомни мне ужесточить правила линта

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    })
+
+    await test.step('Проверяем корректное отображение месяца в селект меню "Month".', async () => {
+      await datePicker.verifyMonthOrYearInSelectMenu("month", `${numberMonth - 1}`)
```
двойные кавычки

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+      await datePicker.verifyDataPickerTabByState('datePickerMonthYear', 'toBeHidden')
+    })
+
+    await test.step(`Сравниваем поле виджета "Select Date".`, async () => {
```
Сравниваем с чем?

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+      await datePicker.selectDayByNumber(day)
+    })
+
+    await test.step(`Проверка изменения цвета ${day}.`, async () => {
```
ты вроде проверяешь что день выбран. Кстати можно попробовать поменять проверку цвета на проверку фокуса, но я не уверен

IlyaKirdun replied:
Я не нашёл чтобы при выборе дня использовался фокус

rybignat commented on tests/widgets/dataPicker.spec.ts:
```diff
@@ -0,0 +1,498 @@
+    })
+  })
+
+  test('CASE_12: Проверка функционала выпадающего меню года в "Date And Time".', async () => {
```
мб я что то не понмаю, но проверка выпадающего меню уже была в 7 кейсе

rybignat reviewed (changes requested):
Требуются правки

IlyaKirdun committed:
Delete .idea/dictionaries directory

IlyaKirdun committed:
Delete .idea/inspectionProfiles directory

IlyaKirdun committed:
Переименовал методы verifyDayStatus(),verifyTimeStatus();
Удалил лишние типы в методе verifyMonthOrYearInDropdownMenu();
Переделал метод getCurrentDateForDateAndTime();
Незначительные исправления.

IlyaKirdun committed:
Перенес тестовые данные в объекты dateAndTimeTestData, selectDateTestData. Переделал тесты с применением их.
Добавил интерфейсы для dateAndTimeTestData, selectDateTestData.
Сделал незначительные исправления.

rybignat commented on page/main.page.ts:
```diff
@@ -8,29 +8,42 @@ export default class MainPage {
 
-    throw new Error('❌ Failed to load the page after 5 attempts.')
+    throw new Error('Не удалось загрузить страницу после 5 попыток.')
   }
```
Вот зачем тебе здесь try catch если у тебя сплошные ифы? Какой смысл?

IlyaKirdun committed:
dataPicker.spec:
1)Переделал импорты под ассоциативные пути;
2)Добавил декораторы в test.beforeEach;
3)Добавил\переделал Pre-conditions в 2,5,10,11 тестах.
Переписал метод navigateToMainPage().
Добавил alias в tsconfig.

IlyaKirdun committed:
Добавил в методе attemptNavigate() выброс ошибки.
Исправил метод navigateToMainPage().

IlyaKirdun committed:
Переработал методы navigateToPageByUrl(),navigateToMainPage().
